lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,148 @@
+
+        self.ppi = PrometheusPushgatewayInterface(self)
+        self.framework.observe(self.ppi.on.pushgateway_available, self._on_pushgateway_available)
+
+    def _on_pushgateway_available(self, _):
+        self._stored.pushgateway_available = True
+```
+"""
+
+import json
+import logging
+from typing import Union
+
+import requests
+from ops.charm import CharmBase, HookEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents, StoredState
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["requests"]
+
+
+class _PushgatewayAvailableEvent(EventBase):
+    """Event emitted when the Prometheus Pushgateway is available to receive metrics."""
+
+
+class _Events(ObjectEvents):
+    """Events for the interface."""
+
+    pushgateway_available = EventSource(_PushgatewayAvailableEvent)
+
+
+class PrometheusPushgatewayInterface(Object):
+    """Interface for the Prometheus Pushgateway."""
+
+    _stored = StoredState()
+    on = _Events()
+
+    def __init__(self, charm: CharmBase):
+        """Construct the interface for the Prometheus Pushgateway.
+
+        Args:
+            charm: a `CharmBase` object that manages this
+                `MetricsEndpointProvider` object. Typically, this is
+                `self` in the instantiating class.
+        """
+        super().__init__(charm, None)
+        self._stored.set_default(pushgateway_url=None)
+
+        self.framework.observe(charm.on.pushgateway_relation_created, self._on_push_relation)
+        self.framework.observe(charm.on.pushgateway_relation_changed, self._on_push_relation)
+
+    def _on_push_relation(self, event: HookEvent) -> None:
+        """Receive the push endpoint information."""
+        raw = event.relation.data[event.app].get("push-endpoint")
+        if raw is not None:
+            logger.info("Received push endpoint information: %r", raw)
+            info = json.loads(raw)
+            self._stored.pushgateway_url = "http://{hostname}:{port}/".format_map(info)
+            self.on.pushgateway_available.emit()
+
+    def send_metric(self, name: str, value: Union[float, int]):
+        """Send a metric to the Pushgateway."""
+        if not isinstance(name, str) or not name.isascii() or not name:
+            raise ValueError("The name must be a non-empty ASCII string.")
+        if not isinstance(value, (float, int)):
+            raise ValueError("The metric value must be an integer or float number.")
+
+        payload = f"{name} {value}\n".encode("ascii")
+        post_url = self._stored.pushgateway_url + "metrics/job/testjob"
+
+        resp = requests.post(post_url, data=payload)
```

Comment:
Would it be easy enough to switch to urllib? Less deps...

Reply:
It could be. But `requests` are also easily testeable with `responses`.

Is there a reason to have few dependencies?

Reply:
Faster `tox --recreate` and `charmcraft pack`.

Reply:
yeah, requests feels like a big dependency to pull in for one post call

Reply:
Done! Now it's all standard library.